Assets/Scripts/KeyLocationController.cs

```diff
@@ -6,55 +6,74 @@
+            foreach (GameObject player in players)
             {
-                float dist = Vector3.Distance(player.transform.position, transform.position);
-                if (dist <= radius && dist > radius - 1)
+                if (player.FindClosestObject("KeyLocation") == gameObject) //Only run if this is the closest keyLocation.
                 {
-                    if (player.transform.position.x < transform.position.x) //Look at what side of the key location the player is at, so we only stop movement in the wanted direction.
-                    {
-                        player.GetComponent<FirstPersonController>().isPlayerKeyLocXPositive = true;
-                    } 
-                    else
+                    float dist = Vector3.Distance(player.transform.position, transform.position);
+
+                    if (!player.GetComponent<FirstPersonController>().isPlayerInKeyLocZone && dist <= radius)
                     {
-                        player.GetComponent<FirstPersonController>().isPlayerKeyLocXPositive = false;
+                        player.GetComponent<FirstPersonController>().isPlayerInKeyLocZone = true;
                     }
 
-                    if (player.transform.position.z < transform.position.z)
+                    if (player.GetComponent<FirstPersonController>().isPlayerInKeyLocZone && dist > radius - 1)
                     {
-                        player.GetComponent<FirstPersonController>().isPlayerKeyLocZPositive = true;
+                        if (player.transform.position.x < transform.position.x) //Look at what side of the key location the player is at, so we only stop movement in the wanted direction.
+                        {
+                            player.GetComponent<FirstPersonController>().isPlayerKeyLocXPositive = true;
+                        }
+                        else
+                        {
+                            player.GetComponent<FirstPersonController>().isPlayerKeyLocXPositive = false;
+                        }
+
+                        if (player.transform.position.z < transform.position.z)
+                        {
+                            player.GetComponent<FirstPersonController>().isPlayerKeyLocZPositive = true;
+                        }
+                        else
+                        {
+                            player.GetComponent<FirstPersonController>().isPlayerKeyLocZPositive = false;
+                        }
+
+                        float t = radius - dist;
+                        speedMod = Mathf.Lerp(-1f, 1, t); //Use difference in distance to key location, and its radius to determine movement speed modifier. Negative values make it so players can allow to be pushed a bit, and still remain stuck as they will rebound to zone edge.
+                        player.GetComponent<FirstPersonController>().keyLocationSpeedMod = speedMod;
```

Comment:
It's been here from before I know, but could you please rename the variables `t` and `dist`? I'm not fully sure what the first one is, but I suspect that `dist` is distance from the player to the border?

Reply:
Sure. t is just the example variable unity uses as the last .Lerp parameter. I'll add something more suited.